docs: add coordinator upgrades design (#5880)
holochain/holochain: docs/design/coordinator_upgrades.md

*both* a new coordinator set and a new integrity version, so `update_app` leaves
a clean **seam** for an integrity-version change to be delegated to chain
continuation rather than forbidding it. Applying an app update is always the
user's choice, never automatic or developer-forced.

## Terminology

- An **integrity zome** defines entry/link types and the validation rules that
  govern them. Integrity code is what is validated against, and it contributes to
  the `DnaHash` (along with the network seed and properties). Changing it is a
  *migration*, not a coordinator upgrade.
- A **coordinator zome** exposes the callable zome functions an app and its UI
  use. Coordinators hold no validation authority and do not affect the
  `DnaHash`; they may be replaced freely, including with coordinators from a
  different publisher. Replacing them does, however, change the zome's public API
  as seen from outside Holochain.
- A **coordinator set** is the collection of coordinator zomes bound to one role
  of an installed app. A cell runs exactly **one** coordinator set at a time —
  the latest installed.
- A **coordinator upgrade** replaces the coordinator set of an installed app with

>   a new one. Integrity code, network, agent key, and source chain are untouched.

Reviewer: An app can have more than one role, so this should be "coordinator sets" plural on upgrade?

Author: Good catch — reworded to plural: a coordinator upgrade 'replaces one or more of an installed app's coordinator sets — an app may have several roles, each with its own set — with new ones.'

1. **It is not app-aware.** An app is the unit a user installs, updates, and
   reasons about. Upgrading coordinators one DNA hash at a time does not match
   how apps are released.
2. **It says nothing about capabilities.** Swapping coordinator code can silently
   change which zome functions exist, orphaning cap grants that referenced them,
   with no defined behaviour.
3. **It cannot co-ordinate a combined release.** A real app update frequently
   changes coordinators *and* integrity together; a per-DNA coordinator swap has
   no place to carry the integrity change.

This design adds an app-level `update_app` as the public path for upgrading an
app, and defines the coordinator and capability lifecycle across an upgrade.
`update_coordinators` leaves the public admin API; any equivalent per-DNA
operation `update_app` needs becomes an internal detail.

## Design decisions

1. **Multiple integrity zomes per DNA are supported.** A DNA may declare more
   than one integrity zome. Entry/link types are namespaced per-zome (an
   `AppEntryDef` carries `(zome_index, entry_index)`) and app-validation dispatch
   routes each op to its defining integrity zome by `zome_index`. An integrity

>    version is therefore the DNA's *set* of integrity zomes.

Reviewer: This is just already how the system is. Do we need to restate it?

Author: Agreed — that's existing behavior, not a decision this design makes. Removed decision 1 and renumbered. The fact that a DNA may hold several integrity zomes is kept only where it's load-bearing (the coordinator-dependency decision and the manifest note).

2. **`network_seed` and `properties` stay in the DNA manifest as defaults,**
   overridable per role via `dna.modifiers` (with a further override at install).
   A different seed/properties choice yields a different `NetworkHash` — a
   different network — so there is no silent divergence of validation rules within

>    one network wherever the value is set.

Reviewer: Is this a change or leaving things as they are?

Author: Leaving things as they are — it's the existing two-layer behavior, not a change. Dropped it from Design decisions (which now lists only actual changes) and kept a one-line 'behaviour left untouched' note plus the manifest Notes.

3. **Coordinators are bound to the app, not the DNA.** The coordinator set is
   declared per role in the app manifest, not inside the DNA bundle. This makes a

>    coordinator a swappable, app-owned thing and is the precondition for app-level

Reviewer: Better word than "thing" please

Author: Reworded — 'a swappable, app-owned component'.

   uses; when there is only one, the dependency is implied and the field may be
   omitted.
5. **`clone_limit` is a role field.** It is a provisioning concern, not a DNA
   property, and lives on the role.
6. **The coordinator declaration is a flat list.** With nothing else expected
   beside the zome list, the coordinator declaration is the list itself rather
   than a `zomes:`-nested object.

## Manifest changes

### DNA manifest — coordinators move out

The DNA manifest loses coordinators (moved to the app). It keeps its
`network_seed`/`properties` defaults and its integrity zome list — a DNA may
still declare several integrity zomes.

Before (today, abbreviated):

```yaml
manifest_version: "0"
name: group
integrity:
  network_seed: 00000000-0000-0000-0000-000000000000
  properties: ~
  zomes:
    - name: group_integrity
      path: ../target/.../group_integrity.wasm
    - name: profiles_integrity
      path: ../target/.../profiles_integrity.wasm
coordinator:
  zomes:
    - name: group
      path: ../target/.../group_coordinator.wasm
      dependencies:
        - name: group_integrity
```

After (proposed):

```yaml

> manifest_version: "1"

Reviewer: This seems like the wrong time to stabilize manifests. Perhaps after DNA migration is done but not as part of this design. Leave the version at 0

Author: Agreed — left both example manifests at manifest_version `"0"` with a comment noting it's not stabilised here. Version bump can wait until after DNA migration.

name: group
integrity:            # seed/properties kept as defaults; only coordinators removed
  network_seed: 00000000-0000-0000-0000-000000000000
  properties: ~
  zomes:              # a DNA may have several integrity zomes
    - name: group_integrity
      path: ../target/.../group_integrity.wasm
      hash: ~         # optional pin (lock/verify); never author-required
    - name: profiles_integrity
      path: ../target/.../profiles_integrity.wasm
# coordinator: block removed — coordinators are declared in the app role
```

### App manifest — role owns the coordinator set and clone limit

The role gains the coordinator set and `clone_limit`; the DNA sub-block keeps
where to find the DNA and its modifier overrides (seed/properties), as today.

Before (today, abbreviated):

```yaml
manifest_version: "0"
roles:
  - name: group
    provisioning:
      strategy: create
      deferred: false
    dna:
      path: ../path/to/group.dna

>       modifiers: { network_seed: ~, properties: ~ }

Reviewer: Flat YAML style please, not embedded JSON

Author: Fixed — expanded the inline flow-map to block YAML for `modifiers`.

      installed_hash: ~
      clone_limit: 5
    # coordinators live inside the DNA bundle, not here
```

After (proposed):

```yaml
manifest_version: "1"
roles:
  - name: group
    provisioning:
      strategy: create
      deferred: false
    clone_limit: 5                     # moved up to the role
    dna:
      path: ../path/to/group.dna
      installed_hash: ~
      modifiers:                       # overrides the DNA's seed/properties defaults
        network_seed: ~
        properties: ~
    coordinators:                      # flattened list, bound to the app
      - name: group
        path: ../target/.../group_coordinator.wasm
        hash: ~                        # optional pin

>         dependencies: [group_integrity]  # which integrity zome's types it uses;

Reviewer: This value is a list too, do we properly support depending on multiple integrity zomes?

Author: No — the platform currently supports at most one integrity-zome dependency per coordinator, even though the field is a list. A coordinator needing types from several integrity zomes in one DNA isn't expressible today. I've stated the current limit honestly and moved 'whether to lift it' to Open questions rather than implying multi-dependency works.

- Seed/properties keep their existing two-layer form: defaults in the DNA
  manifest, overridden by the role's `dna.modifiers`, with a further override at
  install. This layering is unchanged by this design.

## The `update_app` API

```
update_app(installed_app_id, app_bundle) -> AppInfo
```

`update_app` takes an installed app and a new app bundle describing the **complete
desired state** of that app, and reconciles the installation to it.

### Update means strict state

The new bundle is the full intended coordinator/role state, not a delta. There is
no partial "add these coordinators" mode. Strict state means the reconciliation
has a single, well-defined outcome, and orphaned coordinators — present in the
installation but absent from the new bundle — are removed rather than left
dangling.

### Reconciliation algorithm

For each role in the new bundle:

1. **Integrity.** Compute the role's integrity version from the bundle.
   - If it matches the installed integrity version → no migration; proceed to
     coordinators.
   - If it differs → **delegate to the migration path** (chain continuation).
     Until continuation ships, return a `MigrationNotSupported` error naming the

>      role. `update_app` never silently ignores an integrity change.

Reviewer: We should validate the entire update before starting. Not stop per role. That could do a partial update?

Author: Good catch — reworked into two phases: phase 1 validates every role and applies nothing; if any role fails (e.g. an unsupported migration) the whole call errors with the installation untouched. Phase 2 only runs once all roles validate. No partial updates.

2. **Coordinators (strict).** Reconcile the installed coordinator set to the
   bundle's set for this role, by coordinator name:
   - Present in both → **replace** the coordinator WASM under that name.
   - In the bundle only → **install** the new coordinator.
   - Installed only, absent from the bundle → **remove** it.
3. **Capabilities.** Coordinator changes are followed by the capability
   reconciliation described in [Coordinators and
   capabilities](#coordinators-and-capabilities).

> 4. **Upgrade hook.** Each installed/replaced coordinator's upgrade hook is

Reviewer: Should it be lazily invoked like `init` is?

Author: Yes — made it lazy like `init`: `update_app` no longer runs it inline; it's invoked on the first zome call after the upgrade, before any other zome function of the new coordinator.

- To carry a capability forward, the new coordinator **re-creates** the grants it
  wants during its upgrade hook (or at any later time); re-creation stamps them
  with the new coordinator hash. Grant creation is **idempotent**, so a hook can
  safely (re-)create grants without tracking whether a prior version already did.
- A capability check honours a grant only if the grant's coordinator hash matches
  the coordinator currently installed under that name. Grants pointing at
  non-extant coordinators are inert (present but never satisfied).

### Remote calls and claims

- **Remote calls** may optionally specify the coordinator hash to target, for
  callers that need to pin a specific coordinator version; by default a call
  routes to the current coordinator under the named zome. This addresses the
  "zome calls routed to the wrong hApp/coordinator" class of bug
  ([holochain/holochain#2145](https://github.com/holochain/holochain/issues/2145)).
- **Cap claims** are held by the caller and are matched against the grantor's
  current coordinator, not pinned by the claim holder to a grantor hash they
  cannot know.

> ## The upgrade hook

Reviewer: Name the upgrade hook in this section. Is it just `init_upgrade` or something else? Should it have the same "must invoke before other functions" semantics that `init` does?

Author: Named it `init_upgrade` (marked provisional) and gave it `init`'s semantics explicitly: lazy invocation, must complete before any other zome function of the new coordinator runs. Open to a better name if you have one.

A coordinator needs a defined moment to do post-upgrade setup — most importantly,
to **re-create the capability grants** it wants to carry forward (see above). This
design introduces an **upgrade hook** for that purpose, distinct from `init`:

> - **`init` is genesis-only.** It runs once, at genesis, and does not re-run on a

Reviewer: Kind of, but it is invoked lazily so it runs after genesis, but before any other zome functions are allowed to run.

Author: Corrected — the init bullet now says it runs once, lazily: not eager at genesis, but on the first zome call after genesis and before any other zome function is allowed to run.

2. `update_app` reconciles each role: an integrity change is delegated to the
   migration path, and the coordinator set is strict-state reconciled
   (replace/install/remove).
3. The upgrade hook runs and re-creates grants for the new coordinators.

Shipping coordinator upgrades first means step 2's integrity branch returns
`MigrationNotSupported` for now; the coordinator branch is fully functional
standalone. When continuation lands, the same `update_app` surface starts
carrying integrity changes with no change to how developers invoke it.

## Offline friendliness

A coordinator upgrade is a local operation: loading new coordinator WASM,
reconciling the coordinator set, and dispatching the upgrade hook require no
network. Existing data remains readable and its validation is unaffected, since
integrity code did not change. The upgrade hook itself has network access, so any
networked work an app chooses to do there is the app developer's responsibility
(as with `init`).

## Open questions and follow-ups

- **Side-by-side coordinator sets sharing a DNA.** The older draft explored two
  or more coordinator sets over one DNA; this design assumes a single set per

>   role. Multi-set sharing is deferred.

Reviewer: Unless a reason to do something like that comes up, this question won't be picked up. Let's drop it.

Author: Dropped — removed the side-by-side coordinator sets open question.

- **`UseExisting` / cell reuse.** Cross-app cell reuse and the capability
  questions it raises (how one app grants another capability over a shared cell)
  are out of scope here and left to the migration/`UseExisting` work. Note the
  current `CellProvisioning::UseExisting` is already deprecated in favour of
  updating coordinators for late binding and bridge calls for cross-app.
- **Capability detail.** The grant/claim matching rules above are the model, not
  a wire format; exact storage of the injected coordinator hash and the

>   claim-side matching path need a detailed pass.

Reviewer: Then we should do that

Author: Done — moved it out of Open questions and wrote the concrete model into a 'Storage and matching' subsection: the coordinator hash is recorded on the `CapGrant` as a system field at commit time, and the check compares it against the currently-installed coordinator's hash for that zome name. Claim-side stays as-is.

- **Package-management dependency graph.** The older draft's UI→coordinator→
  integrity dependency-graph and lockfile ideas are a packaging concern and are

>   not part of this runtime design; recorded as a possible later addition.

Reviewer: Do they solve a useful problem right now? If not, I suggest we drop this.

Author: No useful problem right now — it's a packaging concern, not runtime. Dropped the package-management dependency-graph bullet.

```

`update_app` takes an installed app and a new app bundle describing the **complete
desired state** of that app, and reconciles the installation to it.

### Update means strict state

The new bundle is the full intended coordinator/role state, not a delta. There is
no partial "add these coordinators" mode. Strict state means the reconciliation
has a single, well-defined outcome, and orphaned coordinators — present in the
installation but absent from the new bundle — are removed rather than left
dangling.

### Reconciliation algorithm

`update_app` **validates the whole update before changing anything**, so it never
leaves an app half-updated. It runs in two phases.

**Phase 1 — validate every role, apply nothing.** For each role in the bundle,
compute the plan without touching the installation:

- **Integrity.** Compare the role's integrity version against the installed one.
  If it differs, the role needs a migration; until chain continuation ships this
  makes the whole update fail with `MigrationNotSupported` naming the role.
  `update_app` never silently ignores an integrity change.
- **Coordinators.** Diff the installed coordinator set against the bundle's set

>   for this role, by coordinator name, into replace/install/remove actions.

Author: Good catch — reworded Phase 1 to compare each role's effective DNA identity (the DnaHash: integrity code together with network seed and properties) against the installed one, rather than an undefined 'integrity version'. Any change to integrity, seed, or properties is now treated as a migration; a coordinator update is never applied to a role whose DNA identity moved. Also updated the combined-update section to say the same.

If *any* role fails validation, `update_app` returns an error and the
installation is untouched — no partial update.

**Phase 2 — apply.** Only once every role validates, apply the computed plans:

1. **Coordinators (strict).** For each role, by coordinator name:
   - Present in both → **replace** the coordinator WASM under that name.
   - In the bundle only → **install** the new coordinator.
   - Installed only, absent from the bundle → **remove** it.
2. **Capabilities.** Coordinator changes are followed by the capability
   reconciliation described in [Coordinators and
   capabilities](#coordinators-and-capabilities).
3. **New / removed roles.** A **new role** registers its DNA and instantiates its
   cell, as at install; a role present only in the installed app is removed under

>    the same strict-state policy.

Author: Addressed — softened the guarantee to match what the phases actually provide, and defined phase-2 atomicity. Phase 1 changes nothing, so it can't leave a partial state; phase 2 now commits the coordinator/capability/DNA-registry writes in a single transaction, with cell instantiation (the one step with external side effects) ordered last and torn down on failure, rolling back to the pre-upgrade state.

>    the same strict-state policy.

Author: Fixed — new roles now follow their manifest provisioning strategy exactly as at install: an immediate role registers its DNA and instantiates its cell, while a provisioning.deferred role registers its DNA but is not instantiated until deferred provisioning.

grant to the coordinator it was created under**, with the coordinator hash
**injected by the system** rather than written by the app.

- A grant is created referencing `(zome_name, function_name[])` as today. The
  system records the hash of the coordinator currently installed under that
  name at creation time: effectively `(coordinator_hash, zome_name,
  function_name[])`. App code neither writes nor needs to know the hash.
- On a coordinator upgrade, grants created under the old coordinator hash **do
  not automatically transfer**. They remain present but point at a coordinator
  that is no longer installed, so the capability check will not honour them.
- To carry a capability forward, the new coordinator **re-creates** the grants it
  wants in its `init_upgrade` hook (or at any later time); re-creation stamps them
  with the new coordinator hash. Grant creation is **idempotent**, so a hook can
  safely (re-)create grants without tracking whether a prior version already did.
- A capability check honours a grant only if the grant's coordinator hash matches
  the coordinator currently installed under that name. Grants pointing at

>   non-extant coordinators are inert (present but never satisfied).

Author: Addressed — added a legacy-matching rule. Grants committed before this design carry no coordinator hash and keep working by matching on (zome_name, function_name[]) alone; the hash check applies only to grants stamped under this design. A coordinator that wants a legacy grant bound to its hash re-creates it (idempotent, same as forward-carried grants).

>   non-extant coordinators are inert (present but never satisfied).

Author: Correction on direction: rather than a legacy-matching/backfill path, we're treating the coordinator-hash field as a breaking change to CapGrant. There will be no existing cap grants when this lands, so no compatibility rule is needed. Updated the doc to state this explicitly and removed the legacy-matching bullet.

### Remote calls and claims

- **Remote calls** may optionally specify the coordinator hash to target, for
  callers that need to pin a specific coordinator version; by default a call
  routes to the current coordinator under the named zome. This addresses the
  "zome calls routed to the wrong hApp/coordinator" class of bug
  ([holochain/holochain#2145](https://github.com/holochain/holochain/issues/2145)).
- **Cap claims** are held by the caller and are matched against the grantor's
  current coordinator, not pinned by the claim holder to a grantor hash they

>   cannot know.

Author: Fixed the contradiction — since a cell runs exactly one coordinator set, there is no older version to target. Remote calls now route to the currently-installed coordinator under the named zome; the optional coordinator hash is a guard (the call proceeds only on a match, else is rejected) rather than a version selector. This still addresses the wrong-hApp routing bug without implying old versions stay executable.

has a single, well-defined outcome, and orphaned coordinators — present in the
installation but absent from the new bundle — are removed rather than left
dangling.

### Reconciliation algorithm

`update_app` **validates the whole update before changing anything**. It runs in
two phases: phase 1 changes nothing, so it cannot leave the app partly updated,
and phase 2 is applied atomically, so a failure mid-apply rolls back to the
pre-upgrade state.

**Phase 1 — validate every role, apply nothing.** For each role in the bundle,
compute the plan without touching the installation:

- **DNA identity.** Compare the role's effective DNA identity — the `DnaHash`,
  which is integrity code together with the network seed and properties — against
  the installed one. Any difference (changed integrity, seed, or properties) means
  the role needs a *migration*; until chain continuation ships this makes the whole
  update fail with `MigrationNotSupported` naming the role. `update_app` never
  silently applies a coordinator update to a role whose DNA identity moved.
- **Coordinators.** Diff the installed coordinator set against the bundle's set

>   for this role, by coordinator name, into replace/install/remove actions.

Author: Fixed — added role-field reconciliation. Phase 1 now diffs the bundle's role-level fields (clone_limit, provisioning) against the installed role: fields that can change in place are planned for application, and a field that can't be reconciled in place fails the role. Phase 2 applies the reconciled role fields so the installed role matches the bundle. Strict state now covers role fields, not just DNA identity and coordinators.

  name at creation time: effectively `(coordinator_hash, zome_name,
  function_name[])`. App code neither writes nor needs to know the hash.
- On a coordinator upgrade, grants created under the old coordinator hash **do
  not automatically transfer**. They remain present but point at a coordinator
  that is no longer installed, so the capability check will not honour them.
- To carry a capability forward, the new coordinator **re-creates** the grants it
  wants in its `init_upgrade` hook (or at any later time); re-creation stamps them
  with the new coordinator hash. Grant creation is **idempotent**, so a hook can
  safely (re-)create grants without tracking whether a prior version already did.
- A capability check honours a grant only if the grant's coordinator hash matches
  the coordinator currently installed under that name. Grants pointing at
  non-extant coordinators are inert (present but never satisfied).
- Grants committed **before this design** carry no recorded coordinator hash.
  They keep working via legacy matching on `(zome_name, function_name[])` alone;
  the hash check applies only to grants stamped under this design. A coordinator

>   that wants a legacy grant bound to its hash re-creates it, as above.

Author: This is resolved by the breaking-change decision (commit 30d4856): there is no legacy hashless grant to migrate. The coordinator hash is a required CapGrant field with no legacy-matching path, and no cap grants will exist when this lands, so there's nothing to revoke, replace, or transition. The re-creation semantics apply only to grants already stamped with a coordinator hash.

# Coordinator Upgrades Design

## Status

**Draft / proposed.** This document describes **coordinator upgrades**: replacing
the coordinator code of an installed app with a new version, at the app level,
without touching the integrity rules, the network, or the source chain.

It modernises an older "Update Coordinators" spec that predated large parts of
the current system (the unified per-DNA database, the v2 model types, and the
[chain continuation](./dna_migration_chain_continuation.md) migration design).

Coordinator upgrades are designed to **compose with, but not depend on**, chain
continuation. They are intended to ship first. An app release will often carry
*both* a new coordinator set and a new integrity version, so `update_app` leaves

> a clean **seam** for an integrity-version change to be delegated to chain

Reviewer 2: Cut it out with your seam, Claude
```suggestion
a clean seam for an integrity-version change to be delegated to chain
```

continuation rather than forbidding it. Applying an app update is always the
user's choice, never automatic or developer-forced.

## Terminology

- An **integrity zome** defines entry/link types and the validation rules that
  govern them. Integrity code is what is validated against, and it contributes to
  the `DnaHash` (along with the network seed and properties). Changing it is a
  *migration*, not a coordinator upgrade.
- A **coordinator zome** exposes the callable zome functions an app and its UI
  use. Coordinators hold no validation authority and do not affect the
  `DnaHash`; they may be replaced freely, including with coordinators from a
  different publisher. Replacing them does, however, change the zome's public API
  as seen from outside Holochain.
- A **coordinator set** is the collection of coordinator zomes bound to one role
  of an installed app. A cell runs exactly **one** coordinator set at a time —
  the latest installed.
- A **coordinator upgrade** replaces one or more of an installed app's coordinator

>   sets — an app may have several roles, each with its own set — with new ones.

Reviewer 2: ```suggestion
- An app may have several roles, each with its own set. A **coordinator upgrade** replaces one or more of an installed app's coordinator set with new ones.
```

The current path is `update_coordinators(dna_hash, coordinator_bundle)`
(`AdminRequest::UpdateCoordinators`). It is a low-level, **per-DNA** operation:
it targets a `DnaHash` directly and knows nothing about apps, roles, or the
capability grants an app relies on. That has three problems:

1. **It is not app-aware.** An app is the unit a user installs, updates, and
   reasons about. Upgrading coordinators one DNA hash at a time does not match
   how apps are released.
2. **It says nothing about capabilities.** Swapping coordinator code can silently
   change which zome functions exist, orphaning cap grants that referenced them,
   with no defined behaviour.
3. **It cannot co-ordinate a combined release.** A real app update frequently
   changes coordinators *and* integrity together; a per-DNA coordinator swap has
   no place to carry the integrity change.

This design adds an app-level `update_app` as the public path for upgrading an
app, and defines the coordinator and capability lifecycle across an upgrade.
`update_coordinators` leaves the public admin API; any equivalent per-DNA
operation `update_app` needs becomes an internal detail.

> ## Design decisions

Reviewer 2: Didn't we also say that the API call `AdminRequest::InstallApp` can no longer override roles and network seed?

Reviewer: We did say that, and partly because I thought it was being introduced. It's existing functionality though. I'm less keen to remove it than I was to avoid adding it in the first place.

I think the network seeds needs to be possible to override. Why roles couldn't be fixed by the packaging I'm not really sure. There's nothing stopping people repackaging an app if they want to.

# Coordinator Upgrades Design

## Status

**Draft / proposed.** This document describes **coordinator upgrades**: replacing
the coordinator code of an installed app with a new version, at the app level,
without touching the integrity rules, the network, or the source chain.

It modernises an older "Update Coordinators" spec that predated large parts of
the current system (the unified per-DNA database, the v2 model types, and the
[chain continuation](./dna_migration_chain_continuation.md) migration design). That
older draft was also internally inconsistent on several points; this document
resolves those, records the decisions taken, and lists what is deliberately left

> open.

Reviewer: ```suggestion
the current system (the unified per-DNA database, the v2 model types, and the
[chain continuation](./dna_migration_chain_continuation.md) migration design).
```

Author: Already in place — the Status paragraph reads exactly this now.

Coordinator upgrades are intended to ship **before** chain continuation and do
**not depend on it**. The two must not conflict, because an app release will
often carry *both* a new coordinator set and a new integrity version. This
document therefore designs coordinator upgrades so that `update_app` leaves a
clean **seam** for an integrity-version change to ride through to chain
continuation, rather than forbidding it (see
[Relationship to chain continuation](#relationship-to-chain-continuation)).

## Terminology

> - An **integrity zome** defines entry/link types and the `validate` rules that

Reviewer: ```suggestion
- An **integrity zome** defines entry/link types and the validation rules that
```

Author: Applied — 'validation rules' instead of '`validate` rules'.

>   govern them. Integrity code is what fixes the `DnaHash` and is validated

Reviewer: It doesn't quite fix the DnaHash because the DnaHash also includes network_seed and DNA properties

Author: Fixed — the bullet now says integrity code 'contributes to the `DnaHash` (along with the network seed and properties)' rather than fixing it alone.

>   against; it changes rarely and, when it does, that is a *migration* (chain

Reviewer: ```suggestion
  against; this design doesn't consider changes to it and, when it does, that is a *migration* (chain
```

More like that? That's less well worded than I'd like but I wouldn't characterize it as rarely necessarily - it's been impossible until now so we don't really know how common it will be.

Author: Applied — dropped the 'changes rarely' characterization; the bullet now just says changing integrity is a migration, not a coordinator upgrade.

  continuation or chain switch), not a coordinator upgrade.
- A **coordinator zome** exposes the callable zome functions an app and its UI
  use. Coordinators hold no validation authority and do not affect the

>   `DnaHash`; they may be replaced freely.

Reviewer: Yes, they may. Even taking coordinators from a different publisher. However, it does affect the public API of the zome as seen from outside of Holochain

Author: Added — coordinators may be replaced freely, including with coordinators from a different publisher, but replacing them changes the zome's public API as seen from outside Holochain.

- A **coordinator set** is the collection of coordinator zomes bound to one role
  of an installed app. A cell runs exactly **one** coordinator set at a time —
  the latest installed.
- A **coordinator upgrade** replaces the coordinator set of an installed app with
  a new one. Integrity code, network, agent key, and source chain are untouched.
- A **DNA migration** changes the integrity rules — via chain continuation (same
  chain, new integrity version) or chain switch (new chain/network). Out of scope

>   here except at the seam.

Reviewer: This isn't relevant then?

Author: Agreed — removed the standalone DNA-migration terminology bullet. Migration is now mentioned only where directly relevant (the seam in `update_app`).

happens: new zome functions, bug fixes, changed call surfaces. None of that
should require a new network, a new chain, or re-validation of existing data —
coordinators do not validate and do not affect the `DnaHash`.

Today the only shipped path is `update_coordinators(dna_hash, coordinator_bundle)`
(`AdminRequest::UpdateCoordinators`). It is a low-level, **per-DNA** operation:
it targets a `DnaHash` directly and knows nothing about apps, roles, or the
capability grants an app relies on. That has three problems:

1. **It is not app-aware.** An app is the unit a user installs, updates, and
   reasons about. Upgrading coordinators one DNA hash at a time does not match
   how apps are released.
2. **It says nothing about capabilities.** Swapping coordinator code can silently
   change which zome functions exist, orphaning cap grants that referenced them,
   with no defined behaviour.
3. **It cannot co-ordinate a combined release.** A real app update frequently
   changes coordinators *and* integrity together; a per-DNA coordinator swap has
   no place to carry the integrity change.

This design adds an app-level `update_app`, defines coordinator/capability

> lifecycle across an upgrade, and keeps `update_coordinators` as the low-level

Reviewer: I think `update_coordinators` should leave the public API even if we have a similar function internally. This seems like the wrong description of what should happen given what I'm after.

Author: Changed — `update_coordinators` now leaves the public admin API; `update_app` is the public path and any per-DNA equivalent it needs is an internal detail. Motivation updated and the 'Relationship to update_coordinators' subsection (which framed it as a retained primitive) is removed.

primitive `update_app` is built on.

## Relationship to chain continuation

Coordinator upgrades and chain continuation are complementary and are designed to
compose into a single app release.

| | Coordinator upgrade | Chain continuation |
|---|---|---|
| Changes | Coordinator zomes only | Integrity version (rules) |
| `DnaHash` / `IntegrityHash` | Unchanged | New `IntegrityHash` |
| Network / chain / key | Unchanged | Network unchanged; chain continues via `ContinueChain` |
| Validation of existing data | Unaffected | Old data validated by old rules |
| Ships | First, standalone | Later |

Chain continuation already establishes two facts this design builds on:

- **One coordinator set, reads every version.** A cell runs a single coordinator
  set (the latest); that coordinator must read data authored under every
  integrity version the chain has used, via app-owned self-tagging enums.
  Coordinator upgrades are how that single set is advanced.
- **Coordinators do not accumulate.** Only integrity versions accumulate in a
  lineage; coordinator zomes are replaced wholesale. So a coordinator upgrade is

>   a *swap*, never an append.

Reviewer: No need to cross reference the integrity behavior everywhere. Just document how to treat coordinators and only mention integrity behavior where it's directly relevant. Otherwise we have two specs to keep in sync for the other feature.

Author: Trimmed — removed the continuation table and the accumulate/one-set facts. Integrity behavior is now mentioned only at the seam, so there aren't two specs to keep in sync.

The seam that keeps them compatible:

- `update_app` accepts an app bundle that may carry a **new integrity version as
  well as** a new coordinator set. Coordinator upgrade handles the coordinator
  swap; the integrity change is **passed through** to the migration path (chain
  continuation) rather than rejected. Until chain continuation lands, `update_app`
  rejects an integrity change with a clear "migration not yet supported" error —
  it never silently drops it, and it never bakes in the assumption that integrity
  *cannot* change (which the older draft did, and which directly contradicts
  continuation).
- The **upgrade hook** introduced here (see [init and the upgrade
  hook](#init-and-the-upgrade-hook)) is the shared place for post-upgrade setup.
  Chain continuation explicitly deferred a dedicated upgrade callback and left
  "seed any content a new integrity version needs" as manual app work; this

>   hook becomes that home too.

Reviewer: Don't document this history, it's twisted. Just define the upgrade hook in this document and why it should exist. Then before we implement the chain continuation, that design can be updated. Not on this PR though.

Author: Removed the history. The upgrade hook is now defined on its own terms — why it exists, distinct from genesis-only `init` — with no continuation-deferral narrative.

## Scope and simplifying decisions

The following decisions shrink the design surface. One proposed simplification —
dropping multiple integrity zomes per DNA — was checked against the current
system and real apps and **rejected**; it is recorded here as decision 1 with the

> evidence, so the choice is not silently revisited.

Reviewer: Don't document the rejected changes

Author: Removed — no more 'considered for removal / rejected' framing. Decisions are stated as outcomes.

   declared per role in the app manifest, not inside the DNA bundle. This is what
   makes a coordinator a swappable, app-owned thing and is the precondition for
   app-level `update_app`.
4. **Keep coordinator `dependencies`, but make it implicit for the single-zome
   case.** Because a DNA may hold several integrity zomes (decision 1), a
   coordinator must still be able to declare which integrity zome's types it
   uses. The `dependencies` field is retained (at most one entry, as today). When
   the DNA has exactly one integrity zome the dependency is implicit and the field
   may be omitted — which is the common case and covers the desired ergonomics
   without dropping the feature multi-zome DNAs need.
5. **Move `clone_limit` to the role.** `clone_limit` is a provisioning concern,
   not a DNA property; it moves from the DNA sub-block to the role.
6. **Flatten the coordinator declaration.** With nothing else expected beside the
   zome list, the coordinator declaration is the list itself rather than a
   `zomes:`-nested object.

## Manifest changes

### DNA manifest — coordinators move out

The DNA manifest loses coordinators (moved to the app). It keeps its

> `network_seed`/`properties` defaults (decision 2) and its integrity zome **list**

Reviewer: We don't need to include the decision making process, just the outcomes as they will be

Author: Reworked into a 'Design decisions' section of plain outcome statements — dropped the evidence dumps and the decision-making narrative.

— seeding any content a new integrity version needs, which continuation left as
manual app work.

## Combined DNA + coordinator update

The target workflow — an app release that changes both — is one `update_app` call:

1. The developer builds a new app bundle whose role carries a new coordinator set
   and (optionally) a new integrity version.
2. `update_app` reconciles each role: an integrity change is passed to chain
   continuation (a `ContinueChain` marker is committed, the new integrity version
   is loaded and recorded in the lineage), and the coordinator set is
   strict-state reconciled (replace/install/remove).
3. The upgrade hook runs: grants are re-created for the new coordinators, and any
   new-version seeding is performed.

Shipping coordinator upgrades first means step 2's integrity branch returns
`MigrationNotSupported` for now; the coordinator branch is fully functional
standalone. When continuation lands, the same `update_app` surface starts
carrying integrity changes with no change to how developers invoke it.

## Offline friendliness

A coordinator upgrade is a local operation: loading new coordinator WASM,

> reconciling the coordinator set, and running the upgrade hook require no network.

Reviewer: The `init` hook has network access available to it right? So in theory this hook should too. That means that whether offline friendliness is up to the app developer.

Author: Right. The upgrade hook has network access like `init`. Reworded: the upgrade-hook section notes it has network access, and Offline friendliness now says whether an upgrade stays fully offline is up to the app developer.

- **Side-by-side coordinator sets sharing a DNA.** The older draft explored two
  or more coordinator sets over one DNA; this design assumes a single set per
  role. Multi-set sharing is deferred.
- **`UseExisting` / cell reuse.** Cross-app cell reuse and the capability
  questions it raises (how one app grants another capability over a shared cell)
  are out of scope here and left to the migration/`UseExisting` work. Note the
  current `CellProvisioning::UseExisting` is already deprecated in favour of
  updating coordinators for late binding and bridge calls for cross-app.
- **Capability detail.** The grant/claim matching rules above are the model, not
  a wire format; exact storage of the injected coordinator hash and the
  claim-side matching path need a detailed pass.
- **Package-management dependency graph.** The older draft's UI→coordinator→
  integrity dependency-graph and lockfile ideas are a packaging concern and are
  not part of this runtime design; recorded as a possible later addition.

## Non-goals

- This document designs coordinator upgrades only. It does not design chain
  continuation or chain switch; it only defines the seam through which an

>   integrity change is delegated to them.

Reviewer: Not really a non-goal, just how I asked you to keep the designs largely independent with conflicts minimized. Don't state here.

Author: Removed. Independence from continuation is now stated as a design principle in Status ('designed to compose with, but not depend on, chain continuation'), not as a non-goal. The Non-goals section is deleted.

- It does not make upgrades automatic or developer-forced. As with migration,

>   applying an app update is the user's choice.

Reviewer: This isn't a non-goal, it's a clear goal that applying updates is a user choice - let's just state that elsewhere and then we don't need a non-goals section at all.

Author: Agreed — moved 'applying an app update is always the user's choice' into Status and deleted the Non-goals section entirely.
